Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp

```diff
@@ -874,6 +787,147 @@ Ref<CoordinatedTileBuffer> CoordinatedPlatformLayer::paint(const IntRect& dirtyR
+        layer.setContentsClippingRect(m_contentsClippingRect);
+
+    if (m_pendingChanges.contains(Change::ContentsBuffer)) {
+        if (m_committedContentsBuffer && m_committedContentsBuffer != m_contentsBuffer)
+            m_committedContentsBuffer->invalidate();
+
+        m_committedContentsBuffer = m_contentsBuffer;
+        if (m_committedContentsBuffer)
+            m_committedContentsBuffer->activateOnCompositingThread(*this);
+    }
+
+    if (m_pendingChanges.contains(Change::ContentsColor))
+        layer.setSolidColor(m_contentsColor);
+
+#if ENABLE(DAMAGE_TRACKING)
+    if (m_pendingChanges.contains(Change::Damage))
+        layer.setDamage(m_damage);
+#endif
+
+    if (m_pendingChanges.contains(Change::Filters))
+        layer.setFilters(m_filters);
+
+    if (m_pendingChanges.contains(Change::Mask))
+        layer.setMaskLayer(m_mask ? &m_mask->ensureTarget() : nullptr);
+
+    if (m_pendingChanges.contains(Change::Replica))
+        layer.setReplicaLayer(m_replica ? &m_replica->ensureTarget() : nullptr);
+
+    if (m_pendingChanges.contains(Change::Backdrop))
+        layer.setBackdropLayer(m_backdrop ? &m_backdrop->ensureTarget() : nullptr);
+
+    if (m_pendingChanges.contains(Change::BackdropRect))
+        layer.setBackdropFiltersRect(m_backdropRect);
+
+    if (m_pendingChanges.contains(Change::Animations))
+        layer.setAnimations(m_animations);
+
+    if (m_pendingChanges.contains(Change::DebugIndicators)) {
+        layer.setShowRepaintCounter(m_repaintCount != -1);
+        layer.setRepaintCount(m_repaintCount);
+
+        layer.setShowDebugBorder(m_debugBorderColor.isVisible());
+        layer.setDebugBorderColor(m_debugBorderColor);
+        layer.setDebugBorderWidth(m_debugBorderWidth);
+    }
+
+    if (m_pendingChanges.contains(Change::Children)) {
+        layer.setChildren(WTF::map(m_children, [](auto& child) {
+            return &child->ensureTarget();
+        }));
+    }
+
+    if (m_backingStoreProxy) {
+        if (!m_backingStore)
+            m_backingStore = CoordinatedBackingStore::create();
+        layer.setBackingStore(m_backingStore.get());
+
+        if (m_animatedBackingStoreClient)
+            layer.setAnimatedBackingStoreClient(m_animatedBackingStoreClient.get());
+
+        auto update = m_backingStoreProxy->takePendingUpdate();
+        m_backingStore->resize(layer.size(), update.scale());
+
+        for (auto tileID : update.tilesToCreate())
+            m_backingStore->createTile(tileID);
+        for (auto tileID : update.tilesToRemove())
+            m_backingStore->removeTile(tileID);
+        for (const auto& tileUpdate : update.tilesToUpdate())
+            m_backingStore->updateTile(tileUpdate.tileID, tileUpdate.dirtyRect, tileUpdate.tileRect, tileUpdate.buffer.copyRef(), { });
+    } else {
+        layer.setBackingStore(nullptr);
+        layer.setAnimatedBackingStoreClient(nullptr);
+        m_backingStore = nullptr;
```

Comment:
I find the m_backingStore and the m_backingStoreProxy, confusing when we set to nullptr. Maybe we can use some function to make sure we encapsulate the update logic of the values? I understand it is dependent but it is in multiple functions.

Reply:
It's indeed confusing, one is used in the composition and should be reset in the compositing thread, the other is the proxy that should be reset in the main thread. I think we should think of new names after all refactorings are done.